graphdatascience/aura_sessions.py

```diff
@@ -0,0 +1,129 @@
+    ) -> None:
+        self._db_credentials = db_credentials
+        self._aura_api = AuraApi(
+            tenant_id=tenant_id, client_id=aura_api_client_auth[0], client_secret=aura_api_client_auth[1]
+        )
+
+    # TODO wrapper around it for closable?
+    def create_gds(self, session_name: str, session_password: str) -> GraphDataScience:
+        if len(session_password) < 8:
+            raise ValueError("Password must be at least 8 characters long.")
+
+        if session_name in [session.name for session in self.list_sessions()]:
+            raise RuntimeError(f"Session with name `{session_name}` already exists.")
+
+        create_details = self._aura_api.create_instance(AuraSessions._instance_name(session_name))
+        self._aura_api.wait_for_instance_running(create_details.id)
+
+        gds_user = create_details.username
+        url = create_details.connection_url
+
+        self._change_initial_pw(
+            gds_url=url, gds_user=gds_user, initial_pw=create_details.password, new_pw=session_password
+        )
+
+        return self._construct_client(gds_url=url, gds_user=gds_user, gds_pw=session_password)
+
+    def connect(self, session_name: str, session_password: str) -> GraphDataScience:
+        instance_name = AuraSessions._instance_name(session_name)
+        matched_instances = [
+            instance.id for instance in self._aura_api.list_instances() if instance.name == instance_name
+        ]
+
+        if len(matched_instances) != 1:
+            raise ValueError("TODO")
+
+        instance_details = self._aura_api.list_instance(matched_instances[0])
+
+        if instance_details:
+            gds_url = instance_details.connection_url
+        else:
+            raise RuntimeError(
+                f"Unable to get connection information for session `{session_name}`. Does it still exist?"
+            )
+
+        # Hardcoded neo4j user as sessions are always created with this user
+        return self._construct_client(gds_url=gds_url, gds_user="neo4j", gds_pw=session_password)
+
+    def delete_gds(self, session_name: str) -> bool:
+        """
+        Delete a GDS session.
+        Args:
+            session_name: the name of the session to delete
+
+        Returns:
+            True iff a session was deleted as a result of this call.
+        """
+        instance_name = AuraSessions._instance_name(session_name)
+
+        candidate_instances = [i for i in self._aura_api.list_instances() if i.name == instance_name]
+
+        if len(candidate_instances) > 1:
+            candidates = [(i.id, i.name) for i in candidate_instances]
+            raise RuntimeError(
+                f"Expected to find exactly one Aura instance with name `{instance_name}`, but found `{candidates}`."
+            )
+
+        if len(candidate_instances) == 1:
+            candidate = candidate_instances[0]
+            return self._aura_api.delete_instance(candidate.id) is not None
+
+        return False
```

Comment:
Isn't this case also an error if finding more than 1 session results in an error? Otherwise I feel like the API here is mixing booleans with exceptions to convey what is going wrong or not.

Reply:
I agree that there is a bit of a mix here. But there is a difference: when there are more than 1 instances this is a bit of an unexpected situation. We cannot forbid our users to rename their AuraDS instances, but we can discourage them from trying to copy our naming pattern. If they still do, this can happen and then it is a situation that arises from bad usage and should result in an error because there's no way to recovery other than manual intervention (e.g. renaming).

But if we find zero sessions that match the name, the situation is a bit different. First, we have the common idempotency characteristic: deleting the same thing twice needn't be an error. If you view the semantics of deletion as "after this operation the resource should not exist" then it is successful behaviour to do nothing if the resource didn't even exist in the first place. We indicate the difference between this and _actually_ deleting in the return value.
Secondly, it is possible that, in a race situation, we successfully find the session instance we should delete, but in the time it takes for us to make the call to the Aura API, the instance was already deleted by another request. Similar to above, we don't need to fail because if it is deleted then we're successful. So we interpret the `404` we will get on the Aura API request as success with a `False`.

Personally, I prefer this design because we don't have to bother our users when they try to delete things that don't exist; compare the `failIfMissing` flag we added to `graph.drop` because users want to have automated scripts that work idempotently.